src/libmongoc/tests/unified/entity-map.c

```diff
@@ -0,0 +1,603 @@
+   event_t *event;
+
+   entity = (entity_t *) mongoc_apm_command_started_get_context (started);
+   event = event_new ("commandStarted");
+   event->command =
+      bson_copy (mongoc_apm_command_started_get_command (started));
+   event->command_name =
+      bson_strdup (mongoc_apm_command_started_get_command_name (started));
+
+   if (should_ignore_event (entity, event)) {
+      event_destroy (event);
+      return;
+   }
+
+   LL_APPEND (entity->events, event);
+}
+
+static void
+command_failed (const mongoc_apm_command_failed_t *failed)
+{
+   entity_t *entity;
+   event_t *event;
+
+   entity = (entity_t *) mongoc_apm_command_failed_get_context (failed);
+   event = event_new ("commandFailed");
+   event->reply = bson_copy (mongoc_apm_command_failed_get_reply (failed));
+   event->command_name =
+      bson_strdup (mongoc_apm_command_failed_get_command_name (failed));
+
+   if (should_ignore_event (entity, event)) {
+      event_destroy (event);
+      return;
+   }
+   LL_APPEND (entity->events, event);
+}
+
+static void
+command_succeeded (const mongoc_apm_command_succeeded_t *succeeded)
+{
+   entity_t *entity;
+   event_t *event;
+
+   entity = (entity_t *) mongoc_apm_command_succeeded_get_context (succeeded);
+   event = event_new ("commandSucceeded");
+   event->reply =
+      bson_copy (mongoc_apm_command_succeeded_get_reply (succeeded));
+   event->command_name =
+      bson_strdup (mongoc_apm_command_succeeded_get_command_name (succeeded));
+
+   if (should_ignore_event (entity, event)) {
+      event_destroy (event);
+      return;
+   }
+   LL_APPEND (entity->events, event);
+}
+
+entity_t *
+entity_client_new (bson_t *bson, bson_error_t *error)
+{
+   bson_parser_t *parser;
+   entity_t *entity;
+   mongoc_client_t *client;
+   mongoc_uri_t *uri = NULL;
+   bool ret = false;
+   bson_iter_t iter;
+   mongoc_apm_callbacks_t *callbacks = NULL;
+   bson_t *uri_options = NULL;
+   bool *use_multiple_mongoses = NULL;
+   bson_t *observe_events = NULL;
+
+   entity = entity_new ("client");
+   parser = bson_parser_new ();
+   bson_parser_utf8 (parser, "id", &entity->id);
+   bson_parser_doc_optional (parser, "uriOptions", &uri_options);
+   bson_parser_bool_optional (
+      parser, "useMultipleMongoses", &use_multiple_mongoses);
+   bson_parser_array_optional (parser, "observeEvents", &observe_events);
+   bson_parser_array_optional (parser,
+                               "ignoreCommandMonitoringEvents",
+                               &entity->ignore_command_monitoring_events);
+
+   if (!bson_parser_parse (parser, bson, error)) {
+      goto done;
+   }
+
+   /* Build the client's URI. */
+   uri = test_framework_get_uri ();
+   if (use_multiple_mongoses && test_framework_is_mongos ()) {
+      /* TODO: Once CDRIVER-3285 is resolved, update this to no longer rely on
+       * the private URI API. */
+      if (*use_multiple_mongoses) {
+         if (!mongoc_uri_upsert_host_and_port (uri, "localhost:27017", error)) {
```

Comment:
will the uri already be populated with a host and port from `test_framework_get_uri()` ? Does it matter, or should that be cleared before upserting the two mongoses?